src/helpers/map.js

```diff
@@ -169,3 +173,150 @@ export function findTile(tiles) {
+ * height: the height of the initialized 2D array
+ * defaultVal: the value to give each element in the initialized 2D array
+ */
+export function init2dArray(width, height, defaultVal) {
+  const matrix = [];
+
+  for (let x = 0; x < width; x++) {
+    matrix[x] = new Array(height);
+    for (let y = 0; y < height; y++) {
+      matrix[x][y] = defaultVal;
+    }
+  }
+
+  return matrix;
+}
+
+
+/*
+ * Returns the sum of all corresponding elements from two matrices being
+ * multiplied together. For example:
+ *   a = 1 2   b = 4 1
+ *       3 4       2 3
+ *   multiplyCorrespondingElementsAndSum(a, b)
+ *   (1*4) + (2*1) + (3*2) + (4*3) = 21
+ *
+ * m1: the first matrix
+ * m2: the second matrix (of the same dimensions as m1)
+ */
+export function multiplyCorrespondingElementsAndSum(m1, m2) {
+  assert(m1.length === m2.length,
+    'multiplyCorrespondingElementsAndSum: m1 and m2 are not the same width');
+  assert(m1[0].length === m2[0].length,
+    'multiplyCorrespondingElementsAndSum: m1 and m2 are not the same height');
+
+  const mWidth = m1.length;
+  const mHeight = m1[0].length;
+  let sum = 0;
+
+  for (let x = 0; x < mWidth; x++) {
+    for (let y = 0; y < mHeight; y++) {
+      sum += m1[x][y] * m2[x][y];
+    }
+  }
+
+  return sum;
+}
+
+
+/* Returns a matrix with a buffer applied around its perimeter. Example:
+ *   Take the element values of the matrix and apply a bufSize perimeter
+ *   of value bufVal.
+ *     mWithBuf = 1 2 3   bufSize = 2   mWithBuf = 1 1 1 1 1 1 1
+ *                4 5 6                            1 1 1 1 1 1 1
+ *                7 8 9   bufVal = 1               1 1 1 2 3 1 1
+ *                                                 1 1 4 5 6 1 1
+ *                                                 1 1 7 8 9 1 1
+ *                                                 1 1 1 1 1 1 1
+ *                                                 1 1 1 1 1 1 1
+ *
+ * m: a 2D array, which will have a buffer added around its perimeter
+ * bufSize: the size of buffer to add around the matrix
+ * bufVal: the value to fill the buffer with
+ */
+export function addBufferTo2dArray(m, bufSize, bufVal) {
+  const mWidth = m.length;
+  const mHeight = m[0].length;
+  const mWithBufWidth = mWidth + (bufSize * 2);
+  const mWithBufHeight = mHeight + (bufSize * 2);
+
+  // Create a grid with matrix's values and a bufSize perimeter of zeros
+  const mWithBuf = init2dArray(mWithBufWidth, mWithBufHeight, bufVal);
+  fillGridWithSubgrid(mWithBuf, m, bufSize, bufSize);
+
+  return mWithBuf;
+}
+
+
+/*
+ * Returns a specified section from a 2D array.
+ *
+ * array: the 2D array to get the subarray from
+ * xCenter: the x index of the center of the subarray
+ * yCenter: the y index of the center of the subarray
+ * width: the width of the subarray (must be an odd number)
+ * height: the height of the subarray (must be an odd number)
+ */
+export function getSubarrayFrom2dArray(array, xCenter, yCenter, width, height) {
```

Comment:
I did think about the level to which we should use `assert`. I tried to catch all of the non-obvious cases that were typically a result of the implementation of the function. For example, `getSubarrayFrom2dArray` could be implemented with non-odd `width` and `height`, but because our use case assumes odd dimensions and it makes the code a bit more intuitive, we applied that constraint.
Are there any places that an `assert` should definitely be added?